### PR TITLE
[.NET] Clean diagnostic rules

### PR DESCRIPTION
**Pull request by paulloz**
Move the following diagnostics into static readonly fields: GD0101, GD0102, GD0103, GD0104, GD0105, GD0106, GD0107, GD0201, GD0202, GD0203, GD0301, GD0302, GD0303, GD0401, GD0402.

To be more consistent, the titles for the following diagnostics were modified: GD0101, GD0105, GD0106, GD0302, GD0303, GD0401, GD0402. A subsequent update of the documentation repo is needed.

Tests for the following diagnostics were created: GD0201, GD0202, GD0203.

### CODE REVIEW

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs [hunk 1]
```diff
@@ -1,7 +1,6 @@
 using System.Linq;
```
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Common.cs:1]: After your changes there should be no usages of LINQ in this file.

```suggestion
```

**Review by raulsntos — COMMENTED** [review]
Thank you so much for this. I think we should merge this first, then update https://github.com/godotengine/godot/pull/87253 and https://github.com/godotengine/godot/pull/88371 to follow the same style.

**Comment by paulloz** [comment]
Applied your comments, and made a few more improvements:
- Always use `'[SomeAttribute]'` instead of `SomeAttribute`
- Always use "type parameter" instead of "type argument"
- Fix the location of GD0401 and GD0402 to not underline the whole class in IDEs

I agree with you for GD0303, but I'm more inclined to retire it in another PR (early in the 4.4 cycle?).

**Comment by raulsntos** [comment]
> Always use "type parameter" instead of "type argument"

These are not the same thing though:

```csharp
class MyType<T> {} // T is a type parameter.
new MyType<string>(); // string is a type argument.
```

So you can only add the `[MustBeVariant]` attribute to a type _parameter_. And you can only check if a type _argument_ is Variant-compatible.

> I agree with you for GD0303, but I'm more inclined to retire it in another PR

Sounds good to me.

**Comment by paulloz** [comment]
> These are not the same thing though

So `GD0301` and `GD0303` should say argument, and `GD0302` should say parameter?

**Comment by raulsntos** [comment]
Yes, `GD0301` should say argument and `GD0302` should say parameter.

I think `GD0303` is a bit more confusing but I think it should say parameter. It's reported when we're unable to find the type parameter that corresponds to a type argument, because we're not handling the kind of symbol that contains the type parameter. But to be honest, I don't really care if we'll be removing it later.

**Comment by paulloz** [comment]
Should be good then.

**Review by raulsntos — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**